#  Analyzing Used Car Listings on eBay Kleinanzeigen

The aim of this project is to clean the data and analyze the included used car listings.

In [1]:
#Import the pandas and NumPy libraries
import pandas as pd
import numpy as np

In [2]:
#Read the autos.csv CSV file into pandas
autos=pd.read_csv("autos.csv",encoding="Latin-1")


In [3]:
autos

dateCrawled  ...             lastSeen
0      2016-03-26 17:47:46  ...  2016-04-06 06:45:54
1      2016-04-04 13:38:56  ...  2016-04-06 14:45:08
2      2016-03-26 18:57:24  ...  2016-04-06 20:15:37
3      2016-03-12 16:58:10  ...  2016-03-15 03:16:28
4      2016-04-01 14:38:50  ...  2016-04-01 14:38:50
...                    ...  ...                  ...
49995  2016-03-27 14:38:19  ...  2016-04-01 13:47:40
49996  2016-03-28 10:50:25  ...  2016-04-02 14:18:02
49997  2016-04-02 14:44:48  ...  2016-04-04 11:47:27
49998  2016-03-08 19:25:42  ...  2016-04-05 16:45:07
49999  2016-03-14 00:42:12  ...  2016-04-06 21:18:48

[50000 rows x 20 columns]

In [4]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled  ...             lastSeen
0  2016-03-26 17:47:46  ...  2016-04-06 06:45:54
1  2016-04-04 13:38:56  ...  2016-04-06 14:45:08
2  2016-03-26 18:57:24  ...  2016-04-06 20:15:37
3  2016-03-12 16:58:10  ...  2016-03-15 03:16:28
4  2016-04-01 14:38:50  ...  2016-04-01 14:38:50

[5 rows x 20 columns]

Our dataset contains 20 columns, most of which are stored as strings. There are a few columns with null values, but no columns have more than ~20% null values. There are some columns that contain dates stored as strings.

We'll start by cleaning the column names to make the data easier to work with.

# Cleaning Column 

In [5]:
#Print an array of the existing column names.
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

The data dictionary provided with data is as follows:

* dateCrawled - When this ad was first crawled. All field-values are taken from this date.
* name - Name of the car.
* seller - Whether the seller is private or a dealer.
* offerType - The type of listing
* price - The price on the ad to sell the car.
* abtest - Whether the listing is included in an A/B test.
* vehicleType - The vehicle Type.
* yearOfRegistration - The year in which the car was first registered.
* gearbox - The transmission type.
* powerPS - The power of the car in PS.
* model - The car model name.
* kilometer - How many kilometers the car has driven.
* monthOfRegistration - The month in which the car was first registered.
* fuelType - What type of fuel the car uses.
* brand - The brand of the car.
* notRepairedDamage - If the car has a damage which is not yet repaired.
* dateCreated - The date on which the eBay listing was created.
* nrOfPictures - The number of pictures in the ad.
* postalCode - The postal code for the location of the vehicle.
* lastSeenOnline - When the crawler saw this ad last online.

In [6]:
# Clean cloumns names:convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive
new_columns_name=['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
autos.columns=new_columns_name    

In [7]:
# Look at the current state of the autos dataframe.
autos.head()

date_crawled  ...            last_seen
0  2016-03-26 17:47:46  ...  2016-04-06 06:45:54
1  2016-04-04 13:38:56  ...  2016-04-06 14:45:08
2  2016-03-26 18:57:24  ...  2016-04-06 20:15:37
3  2016-03-12 16:58:10  ...  2016-03-15 03:16:28
4  2016-04-01 14:38:50  ...  2016-04-01 14:38:50

[5 rows x 20 columns]

# Initial Exploration and Cleaning

Now let's do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for: - Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.

In [8]:
#Look at descriptive statistics for all columns.
autos.describe(include="all")

date_crawled         name  ...   postal_code            last_seen
count                 50000        50000  ...  50000.000000                50000
unique                48213        38754  ...           NaN                39481
top     2016-03-12 16:06:22  Ford_Fiesta  ...           NaN  2016-04-07 06:17:27
freq                      3           78  ...           NaN                    8
mean                    NaN          NaN  ...  50813.627300                  NaN
std                     NaN          NaN  ...  25779.747957                  NaN
min                     NaN          NaN  ...   1067.000000                  NaN
25%                     NaN          NaN  ...  30451.000000                  NaN
50%                     NaN          NaN  ...  49577.000000                  NaN
75%                     NaN          NaN  ...  71540.000000                  NaN
max                     NaN          NaN  ...  99998.000000                  NaN

[11 rows x 20 columns]


Our initial observations:

* There are a number of text columns where all (or nearly all) of the values are the same:
  * seller
  * offer_type
* The num_photos column looks odd, we'll need to investigate this further.

In [9]:
autos["num_photos"].value_counts()

0    50000
Name: num_photos, dtype: int64

It looks like the num_photos column has 0 for every column. We'll drop this column, plus the other two we noted as mostly one value.

In [10]:
#Drop these columns
autos=autos.drop(["num_photos", "seller", "offer_type"],axis=1)

There are two columns, price and auto, which are numeric values with extra characters being stored as text. We'll clean and convert these.

In [11]:
#Clean price column
autos["price"]=autos["price"].str.replace("$","").str.replace(",","").astype(int)
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [12]:
#Clean odometer column
autos["odometer"]=(autos["odometer"].str.replace(",","").str.replace("km","").astype(int))
autos.rename({"odometer":"odometer_km"},axis=1,inplace=True)
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

# Exploring the Odometer and Price columns

In [13]:
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

We can see that the values in this field are rounded, which might indicate that sellers had to choose from pre-set options for this field. Additionally, there are more high mileage than low mileage vehicles.

In [14]:
print(autos["price"].unique().shape)
print(autos["price"].describe())
autos["price"].value_counts().head(20)

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: price, dtype: int64

Again, the prices in this column seem rounded, however given there are 2357 unique values in the column, that may just be people's tendency to round prices on the site.

There are 1,421 cars listed with $0 price - given that this is only 2% of the of the cars, we might consider removing these rows. The maximum price is one hundred million dollars, which seems a lot, let's look at the highest prices further.

In [15]:
#Look at price column in descending order
autos["price"].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [16]:
##Look at price column in ascending order
autos["price"].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64


There are a number of listings with prices below \$30, including about 1,500 at \$0. There are also a small number of listings with very high values, including 14 at around or over $1 million.

Given that eBay is an auction site, there could legitimately be items where the opening bid is \$1. We will keep the \$1 items, but remove anything above \$350,000, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.

In [17]:
#Chose the price between 1 and 350000
autos=autos[autos["price"].between(1,351000)]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

# Exploring the date columns

In [18]:
#Look at these columns
autos[["date_crawled","ad_created","last_seen"]][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [23]:
#calculate the distribution of values in the date_crawled
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [24]:
#calculate the distribution of values in the ad_created
(autos["ad_created"]
         .str[0:10]
         .value_counts(normalize=True,dropna=False)
         .sort_index()
)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

In [25]:
#calculate the distribution of values in the last_seen
(autos["last_seen"]
        .str[0:10]
        .value_counts(normalize=True,dropna=False)
        .sort_index()
)

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

In [26]:
#Describe registration_year column
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

# Dealing with Incorrect Registration Year Data

In [29]:
#Finding what percentage of our data has invalid values in this column:
(~autos["registration_year"].between(1900,2016)).sum()/autos.shape[0]

0.0

In [30]:
#Remove the values outside those upper and lower bounds and calculate the distribution of the remaining values
autos=autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: registration_year, dtype: float64

# Exploring Price by Brand

In [31]:
#Look at brand column in percentage
autos["brand"].value_counts(normalize=True)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

In [34]:
# Finding brands representing more than 5% of total listings.
brand_counts=autos["brand"].value_counts(normalize=True)
common_brand=brand_counts[brand_counts >0.05].index
print(common_brand)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [35]:
# Finding mean price of top 5 brands
brand_mean_price={}

for brand in common_brand:
    mean_price=autos.loc[autos["brand"]==brand,"price"].mean()
    brand_mean_price[brand]=int(mean_price)
    
print(brand_mean_price)

{'volkswagen': 5402, 'bmw': 8332, 'opel': 2975, 'mercedes_benz': 8628, 'audi': 9336, 'ford': 3749}


# Exploring Mileage

In [41]:
#Storing Aggregate data in a Dataframe(first to a series object)
bmp_series=pd.Series(brand_mean_price)
pd.DataFrame(bmp_series,columns=["mean_price"])

mean_price
volkswagen           5402
bmw                  8332
opel                 2975
mercedes_benz        8628
audi                 9336
ford                 3749

In [42]:
# Finding mean mileage  of top 5 brands
brand_mean_mileage={}

for brand in common_brand:
    brand_bool=autos["brand"]==brand
    mean_mileage=autos.loc[brand_bool,"odometer_km"].mean()
    brand_mean_mileage[brand]=int(mean_mileage)
    
print(brand_mean_mileage)

{'volkswagen': 128707, 'bmw': 132572, 'opel': 129310, 'mercedes_benz': 130788, 'audi': 129157, 'ford': 124266}


In [43]:
mean_mileage=pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_price=pd.Series(brand_mean_price).sort_values(ascending=False)


In [44]:
#Storing Aggregate data in a Dataframe
brand_info=pd.DataFrame(mean_mileage,columns=["mean_mileage"])
brand_info

mean_mileage
bmw                  132572
mercedes_benz        130788
opel                 129310
audi                 129157
volkswagen           128707
ford                 124266

In [45]:
#Combine the mean_price
brand_info["mean_price"]=mean_price
brand_info

mean_mileage  mean_price
bmw                  132572        8332
mercedes_benz        130788        8628
opel                 129310        2975
audi                 129157        9336
volkswagen           128707        5402
ford                 124266        3749